src/librustc/traits/select.rs

```diff
@@ -781,9 +781,7 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         debug!("evaluate_trait_predicate_recursively({:?})", obligation);
 
         if self.intercrate.is_none() && obligation.is_global()
-            && obligation.param_env.caller_bounds.iter().all(|bound| {
-                bound.needs_infer() || bound.has_param_types() || bound.has_self_ty()
-            })
+            && obligation.param_env.caller_bounds.iter().all(|bound| bound.needs_subst())
```

Comment:
Hmm, I think that `is_global` *feels* like the right flag here. I suspect that it should potentially just not be false for `ReLateBound`. This will also cause fewer obligations to be cached than I think we *could* cache -- that is, if we have proven something like `for<'a> u32: Trait<'a>`, it will not count for the purpose of this test, but I think it could.

Reply:
We could also modify `is_global` to something like:

```rust
    fn is_global(&self) -> bool {
        !self.has_type_flags(TypeFlags::HAS_LOCAL_NAMES) && !self.has_escaping_regions()
    }
```

or even try just asserting that `has_escaping_regions` is true. This would ensure that any late-bound regions are bound by some `for<'a>` binder when we test things if they are global (which I believe ought to be true)